File: 98-validate-binary-search-tree/validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool fun(TreeNode* node,long minVal,long maxVal)
    {
        if(node==nullptr)
        {
            return true;
        }
        if(node->val <= minVal || node->val>=maxVal)
        {
            return false;
        }
        return fun(node->left,minVal,node->val) && fun(node->right,node->val,maxVal);
    }
    bool isValidBST(TreeNode* root) {
        return fun(root,LONG_MIN,LONG_MAX);
    }
};

File: 98-validate-binary-search-tree/validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    List<Integer> ls = new ArrayList<>();
    private void traverse(TreeNode root)
    {
        if(root==null)
        {
            return;
        }
        traverse(root.left);
        ls.add(root.val);
        traverse(root.right);

        return;
    }
    public boolean isValidBST(TreeNode root) {
        ls.clear();
        traverse(root);

        int n=ls.size();
        for(int i=1;i<n;i++)
        {
            if(ls.get(i)<=ls.get(i-1))
            {
                return false;
            }
        }
        return true;
    }
}